### PR TITLE
envoy 1.21.0

### PR DESCRIPTION
**Pull request by XuehaiPan**
Migrate stand-alone formula `envoy` to Python 3.10. Part of PR #90716.

### CODE REVIEW

**Comment by cclauss** [comment]
[These tests are being run on Python 3.5](https://github.com/Homebrew/homebrew-core/runs/4485370918?check_suite_focus=true#step:7:185) which thinks that Python type hints are a syntax error.

**Comment by branchvincent** [comment]
```
File "/tmp/envoy-20211216-11802-y4loia/.brew_home/_bazel/c8017d5c263d650d08e7b606fae0ff95/sandbox/processwrapper-sandbox/2817/execroot/envoy/external/com_googlesource_chromium_v8/wee8/third_party/jinja2/tests.py", line 13, in <module>
      from collections import Mapping
  ImportError: cannot import name 'Mapping' from 'collections' 
```

**Comment by XuehaiPan** [comment]
It is issued with the Google V8 Engine:

https://github.com/envoyproxy/envoy/blob/8618d52d9244f62f6931f766faee0d9061f2501d/bazel/repository_locations.bzl#L826-L840

```console
$ wget -O - https://storage.googleapis.com/envoyproxy-wee8/wee8-9.6.180.12.tar.gz | tar -xz

$ grep -r 'from collections import Mapping' wee8
wee8/third_party/jinja2/tests.py:from collections import Mapping
wee8/third_party/jinja2/sandbox.py:from collections import Mapping
wee8/third_party/jinja2/runtime.py:    from collections import Mapping
```

`collections.Mapping` was deprecated since Python 3.3 and was removed in Python 3.10. It should change to `from collections.abc import Mapping`.

**Comment by cclauss** [comment]
This might be better as an Envoy issue.  https://github.com/envoyproxy/envoy/issues

**Comment by XuehaiPan** [comment]
Waiting for pull request envoyproxy/envoy#19244 to be merged.

**Comment by XuehaiPan** [comment]
Waiting for a new release of `envoy`.

**Comment by cclauss** [comment]
https://github.com/envoyproxy/envoy/releases

**Comment by XuehaiPan** [comment]
In:
 https://github.com/envoyproxy/envoy/blob/8cdfe788027127210beb08ad58bbc8754c71af2d/source/extensions/transport_sockets/tcp_stats/tcp_stats.cc#L12

```cc
#include </usr/include/linux/tcp.h>
```

It's a hardcoded header path.

------

On Ubuntu 16.04 LTS:

```cc
struct tcp_info {
	__u8	tcpi_state;
	__u8	tcpi_ca_state;
	__u8	tcpi_retransmits;
	__u8	tcpi_probes;
	__u8	tcpi_backoff;
	__u8	tcpi_options;
	__u8	tcpi_snd_wscale : 4, tcpi_rcv_wscale : 4;

	__u32	tcpi_rto;
	__u32	tcpi_ato;
	__u32	tcpi_snd_mss;
	__u32	tcpi_rcv_mss;

	__u32	tcpi_unacked;
	__u32	tcpi_sacked;
	__u32	tcpi_lost;
	__u32	tcpi_retrans;
	__u32	tcpi_fackets;

	/* Times. */
	__u32	tcpi_last_data_sent;
	__u32	tcpi_last_ack_sent;     /* Not remembered, sorry. */
	__u32	tcpi_last_data_recv;
	__u32	tcpi_last_ack_recv;

	/* Metrics. */
	__u32	tcpi_pmtu;
	__u32	tcpi_rcv_ssthresh;
	__u32	tcpi_rtt;
	__u32	tcpi_rttvar;
	__u32	tcpi_snd_ssthresh;
	__u32	tcpi_snd_cwnd;
	__u32	tcpi_advmss;
	__u32	tcpi_reordering;

	__u32	tcpi_rcv_rtt;
	__u32	tcpi_rcv_space;

	__u32	tcpi_total_retrans;

	__u64	tcpi_pacing_rate;
	__u64	tcpi_max_pacing_rate;
	__u64	tcpi_bytes_acked;    /* RFC4898 tcpEStatsAppHCThruOctetsAcked */
	__u64	tcpi_bytes_received; /* RFC4898 tcpEStatsAppHCThruOctetsReceived */
	__u32	tcpi_segs_out;	     /* RFC4898 tcpEStatsPerfSegsOut */
	__u32	tcpi_segs_in;	     /* RFC4898 tcpEStatsPerfSegsIn */
};
```

On Ubuntu 20.04 LTS:

```cc
struct tcp_info {
	__u8	tcpi_state;
	__u8	tcpi_ca_state;
	__u8	tcpi_retransmits;
	__u8	tcpi_probes;
	__u8	tcpi_backoff;
	__u8	tcpi_options;
	__u8	tcpi_snd_wscale : 4, tcpi_rcv_wscale : 4;
	__u8	tcpi_delivery_rate_app_limited:1;

	__u32	tcpi_rto;
	__u32	tcpi_ato;
	__u32	tcpi_snd_mss;
	__u32	tcpi_rcv_mss;

	__u32	tcpi_unacked;
	__u32	tcpi_sacked;
	__u32	tcpi_lost;
	__u32	tcpi_retrans;
	__u32	tcpi_fackets;

	/* Times. */
	__u32	tcpi_last_data_sent;
	__u32	tcpi_last_ack_sent;     /* Not remembered, sorry. */
	__u32	tcpi_last_data_recv;
	__u32	tcpi_last_ack_recv;

	/* Metrics. */
	__u32	tcpi_pmtu;
	__u32	tcpi_rcv_ssthresh;
	__u32	tcpi_rtt;
	__u32	tcpi_rttvar;
	__u32	tcpi_snd_ssthresh;
	__u32	tcpi_snd_cwnd;
	__u32	tcpi_advmss;
	__u32	tcpi_reordering;

	__u32	tcpi_rcv_rtt;
	__u32	tcpi_rcv_space;

	__u32	tcpi_total_retrans;

	__u64	tcpi_pacing_rate;
	__u64	tcpi_max_pacing_rate;
	__u64	tcpi_bytes_acked;    /* RFC4898 tcpEStatsAppHCThruOctetsAcked */
	__u64	tcpi_bytes_received; /* RFC4898 tcpEStatsAppHCThruOctetsReceived */
	__u32	tcpi_segs_out;	     /* RFC4898 tcpEStatsPerfSegsOut */
	__u32	tcpi_segs_in;	     /* RFC4898 tcpEStatsPerfSegsIn */

	__u32	tcpi_notsent_bytes;
	__u32	tcpi_min_rtt;
	__u32	tcpi_data_segs_in;	/* RFC4898 tcpEStatsDataSegsIn */
	__u32	tcpi_data_segs_out;	/* RFC4898 tcpEStatsDataSegsOut */

	__u64   tcpi_delivery_rate;

	__u64	tcpi_busy_time;      /* Time (usec) busy sending data */
	__u64	tcpi_rwnd_limited;   /* Time (usec) limited by receive window */
	__u64	tcpi_sndbuf_limited; /* Time (usec) limited by send buffer */

	__u32	tcpi_delivered;
	__u32	tcpi_delivered_ce;

	__u64	tcpi_bytes_sent;     /* RFC4898 tcpEStatsPerfHCDataOctetsOut */
	__u64	tcpi_bytes_retrans;  /* RFC4898 tcpEStatsPerfOctetsRetrans */
	__u32	tcpi_dsack_dups;     /* RFC4898 tcpEStatsStackDSACKDups */
	__u32	tcpi_reord_seen;     /* reordering events seen */

	__u32	tcpi_rcv_ooopack;    /* Out-of-order packets received */

	__u32	tcpi_snd_wnd;	     /* peer's advertised receive window after
				      * scaling (bytes)
				      */
};
```

Diff:

```diff
8a9
> 	__u8	tcpi_delivery_rate_app_limited:1;
47a49,73
> 
> 	__u32	tcpi_notsent_bytes;
> 	__u32	tcpi_min_rtt;
> 	__u32	tcpi_data_segs_in;	/* RFC4898 tcpEStatsDataSegsIn */
> 	__u32	tcpi_data_segs_out;	/* RFC4898 tcpEStatsDataSegsOut */
> 
> 	__u64   tcpi_delivery_rate;
> 
> 	__u64	tcpi_busy_time;      /* Time (usec) busy sending data */
> 	__u64	tcpi_rwnd_limited;   /* Time (usec) limited by receive window */
> 	__u64	tcpi_sndbuf_limited; /* Time (usec) limited by send buffer */
> 
> 	__u32	tcpi_delivered;
> 	__u32	tcpi_delivered_ce;
> 
> 	__u64	tcpi_bytes_sent;     /* RFC4898 tcpEStatsPerfHCDataOctetsOut */
> 	__u64	tcpi_bytes_retrans;  /* RFC4898 tcpEStatsPerfOctetsRetrans */
> 	__u32	tcpi_dsack_dups;     /* RFC4898 tcpEStatsStackDSACKDups */
> 	__u32	tcpi_reord_seen;     /* reordering events seen */
> 
> 	__u32	tcpi_rcv_ooopack;    /* Out-of-order packets received */
> 
> 	__u32	tcpi_snd_wnd;	     /* peer's advertised receive window after
> 				      * scaling (bytes)
> 				      */
```

File: Formula/envoy.rb [hunk 1]
```diff
@@ -60,6 +60,13 @@ def install
       --host_action_env=PATH=#{env_path}
     ]
 
+    if OS.linux? && OS.kernel_version < "4.6"
+      # Extension `tcp_stats` requires Linux headers >= 4.6
+      # (It's included by absolute path `#include </usr/include/linux/tcp.h>`)
+      # Disable it when the kernel version requirements not satisfied
+      args << "--//source/extensions/transport_sockets/tcp_stats:enabled=false"
+    end
+
```
Comment (XuehaiPan) [inline Formula/envoy.rb:69]: The CI container failed to detect the correct kernel version (Linux Kernel for Ubuntu 16.04 is 4.4.0).

Comment (carlocab) [reply Formula/envoy.rb:69]: My guess is yes. But CC @Homebrew/linux

Comment (iMichka) [reply Formula/envoy.rb:69]: Instead of looking at the kernel version, can we read the `tcp_stats.cc` file and make a decision based on that?

Comment (XuehaiPan) [reply Formula/envoy.rb:69]: https://github.com/envoyproxy/envoy/blob/release/v1.21/source/extensions/transport_sockets/tcp_stats/tcp_stats.cc#L12

**Comment by XuehaiPan** [comment]
Need label `CI-long-timeout`.

**Comment by XuehaiPan** [comment]
All tests passed.

**Review by SMillerDev — APPROVED** [review]

**Comment by XuehaiPan** [comment]
May need to remove the label `automerge-skip`.

**Comment by branchvincent** [comment]
@XuehaiPan thanks for sticking with this one!

**Comment by codefromthecrypt** [comment]
Thanks, folks. This is great progress

**Pull request closed without merging**